designs/pluggable-message-routing/README.md

```diff
@@ -0,0 +1,263 @@
+
+Note: The method for managing and distributing instances running in different cores
+or machines is outside the scope of this design.
+
+
+### High-level design
+
+There are two possible models to use for the pluggable message routing feature.
+
+1. a configurable stack of router layers
+2. a set of life-cycle hooks that can have handlers attached to them
+
+**At this stage, the life-cycle hooks approach is the preferred model.**
+
+#### Stack of layers
+
+The working assumption for this feature has always been to take an approach similar
+to that of Express middleware:
+
+Each message is passed to the top layer of the stack. That layer can do whatever
+it wants with the message before either disposing of it or passing it on to the
+next layer in the stack.
+
+Node-RED would provide a core `LocalRouter` layer that knows how to route messages
+to nodes in the local runtime.
+
+The stack layers could be customised via either the settings file or via runtime API.
+
+The problem with this model comes when you have unrelated things wanting to customise
+the stack.
+
+For example, if a user provided a custom layer to add additional logging, they may
+specify a stack that looks like:
+```
+1. MyCustomLogger
+2. RED.router.LocalRouter
+```
+
+They then enable the Flow Debugger that has to insert itself into this stack. That would
+mean the Flow Debugger would have to keep track of the old stack configuration and be
+able to restore it when the Debugger is disabled.
+
+This could get very complicated to manage.
+
+
+#### Life-Cycle Hooks
+
+This is similar to the Fastify model of registering hooks at certain key points
+in the life-cycle of a message.
+
+Rather than leave the stack as being completely customisable, we recognise there
+is a core set of steps that every message has to go through. Within those steps
+are points where some custom code *might* want to run.
+
+The following diagram shows a possible set of hook points. This list may well get
+refined as the design progresses. The orange lines show the span of synchronous calls.
+
+![](message-router-events.png)
+
+1. `preSend` - passed an array of `SendEvent` objects. The messages inside these objects
+   are exactly what the node has passed to `node.send` - meaning there could be duplicate
+   references to the same message object.
+
+2. `preRoute` - called once for each SendEvent object in turn
+
+3. `onSend` - the local router has identified the node it is going to send to. At
+   this point, the message has been cloned if needed.
+
+4. `postSend` - the message has been dispatched to be delivered asynchronously
+(unless the sync delivery flag is set, in which case it would be continue as synchronous delivery)
+
+5. `onReceive` - a node is about to receive a message
+
+6. `postReceive` - the message has been passed to the node's `input` handler
+
+7. `onDone`, `onError` - the node has completed with a message or logged an error
+
```

Comment:
Do there need to be anything re onDeploy and onClose also for the hook handlers to get hold of ? eg to initialise/close a remote connection ?

Reply:
I haven't got to that part of the lifecycle yet :) But yes, this is opening up a broader hook-based model in the runtime.

With the old Router model, yes the router layers would get notified whenever the flow configuration changed. But it then left the layer to do all its own parsing and figuring out what's what in the flow. And many layers wouldn't care - an enhanced logger for example wouldn't need to do anything, it just wants the message events. I think this is where a more standard Flow API would be helpful (which may pull in the flow manipulation API piece from the flow linter work).

But yes, this is one of the next areas to look at in the design.